### DISCUSSION TITLE
Smart Meter Interface only provides garbage on serial from P1 port

### DISCUSSION THREAD

**Discussion by kobuki** [arendst/Tasmota#22348] [Support request] [answered]
I put together a test gear for reading our new smart meter, a Sagemcom MA105 (single phase). I hooked the P1 port output to GPIO21, configured the meter and a single metric for tests using the following script:

```
>D
>B
->sensor53 r
>M 1
+1,21,o,16,115200,ma105
1,1-0:32.7.0(@1,L1 voltage,V,voltage_l1,17
#
```

But whatever config I try to change, I only get garbage:

```
...
16:42:55.865 : I%M=&sfc24gf4fMM=ocf)V
16:42:55.903 : KM=gifL4d4fMfff&fMI%M=gif34d4fMffsfffMI%M=gifL4d4fMffsffMI$2$M=gifLffsfMfsfffMI$2$M=gifLfffMffsfffMI$2$M=gif
16:42:55.907 : MffffffsfffMI$2dIMMffffffsfffMIO#IMMffffffsfffMI$2dIMMffffffsffMI$2dIMMffffffsfffMI$2dIMMfffffsfffMIeI
16:43:05.803 : MffsfffMI%MMffsfffMIkMffsfffMI%MMffsfffMI%MMffsfffM	TkMffsfffMIEM=&sfLL4gf4fMM=o/]L=
16:43:05.900 : y=&sfLLfSfsfM&L&Lf&&fgfLf&f&f&&&M=&sfLL4g4fMfff&M=&sfLLf4gfM&M=&sfL&fsfsfMf4fffMI%M=gifLfsFx^>`fffLfS.2VR"^ZJV:J>
16:43:05.904 : y=gifLffsfMffsfffMIO#ky=gifLLfsfMffSFL5J"O#ky=gifL&fsfsfMfFsffMiSV>
16:43:05.910 : 3ffL&fffff&%MMffffffsfffMIeIMMffffffsfffM	TIMMffffffsfffMIeIMMffffffsfffM	TIMMffffffsfffMIeIMMfffffsfffMIeIMMffffffsfffMI$
16:43:05.914 : MffffffsfffMI$2dIMMfffffsfffMI$2dIMMffffffsfffMI$2d	kMffffffsfffMI$2dIMMffffffsfffM	"O#IMMffffffsfffMIeIMMffsfffMI%
...
```

What could be the problem? I used direct connection without inverter for the port, since to the best of my knowledge, the Smart Meter Interface inverts the P1 serial output signal on its own. Is that true? I also tried GPIO3 (HW serial) with the same results. Using the new port serial signal inversion feature of 14.3.0 (`SetOption69 1`) I also tried to test the data stream via simple serial logging but it rarely picks up data so something must be amiss.

Any ideas?

**Comment by Pietjuhbel** [comment]
What is your serial config and baudrate?

**Reply by kobuki** [reply]
It's also in the script above, but it's 115200 and 8N1 (default).

**Reply by Pietjuhbel** [reply]
Sorry, overlooked that.
Some p1 interfaces are 9600 baud, perhaps give that a try.

**Reply by kobuki** [reply]
I tried 9600 and othe baud rates with the same results. Actually, it spews out fewer garbage per 10-sec report period, that can indicate that the actual baud rate is higher. It might be needed to invert the signal using additional HW, after all...

**Comment by kobuki** [comment]
It seems I still need the inverter hardware. I've built the simple transistor-based one and now I get meaningful values over the P1 port. It's a shame, though, as the ESP32 is capable of using inverted HW serial and SW serial could also be programmed so.

I've noticed 2 weird things, though.
 - When we export power to the utility from the solar inverter, the power factor is always above 4.0. It spans some low 5.x values as well. When we only import after sunny hours, it normalizes to between 0.0-1.0 as it should (though I will need some more investigation to see that's indeed the deciding factor). What gives?
 - The serial output comes in 10-second intervals. The whole set of telegrams are transmitted in a single burst. But one or more telegrams almost always contain some garbage characters, or a single telegram is missing or truncated. Is this normal? A friend of mine has a Sangxing 3-phase inverter with the same signal-inverting circuit I made for him and he doesn't seem to receive any garbage. Any ideas?

**Comment by kobuki** [comment, marked as answer]
For whomever wanders around here by means of searching for info. The signal inverting hardware is NOT needed, at least as of 14.3.0 and ESP32 (possibly works for other versions and ESP826x, but I haven't tried that). Simply add a line before the metrics definitions that contains `1,=so2,4`. This inverts the HW serial input on GPIO3. It's documented in the special commands section but it escaped my attention earlier.

I don't receive occasional garbage using this method. Now I just directly connect GND, output enable and the data line to the board (I'm not powering the ESP32 from the P1 port). All on 3.3V, I don't connect 5V to the output enable on the P1 port.

The power factor being a weirdly high value when exporting remains a mystery.
